### ISSUE TITLE
Can vroom be used with google maps

### ISSUE THREAD

**Issue by AhtashamImtiyaz** [VROOM-Project/vroom#384]
Setting up osrm locally requires alot of resources. Can vroom be used with google map or is there any paid solution (osrm or openrouteservice) with which vroom can be used?

**Comment by nilsnolde** [comment]
You can plug in any matrix you want. OSRM & ORS are merely for convenience.

https://github.com/VROOM-Project/vroom/blob/master/docs/API.md#matrix

**Comment by AhtashamImtiyaz** [comment]
I tried with google map distance matrix api. The matrix object vrooms require is not possible with google maps

**Comment by nilsnolde** [comment]
Didn't use google maps API in ages, but there's no problem using it with vroom. Of course you'll have to parse and alter the matrix from google to fit into vroom's requirements, but that's a pure data massaging task.

**Comment by jcoupey** [comment]
> I guess the matrix object (array of arrays of time-distance matrices) vrooms require is not possible with google maps

Not sure what you mean exactly here. You have the ability to provide any matrix describing travel times (see documentation link above), whatever the original provider. But in that case you have to first handle the matrix request yourself, then feed the matrix to `vroom` along with indexes for all locations in the problem (see `location_index`, `start_index` keys). See [this example](https://github.com/VROOM-Project/vroom/blob/master/docs/API.md#using-a-custom-matrix) for more.

There is no paid routing service that would currently work out-of-the-box but using any commercial service based on OSRM or OpenRouteService should not be that hard, the main restriction being the ability to set credentials, see #219.

On the other hand, if you want to use optimization without the hassle of deploying your own stack, you may want to go for a paid optimization service. For what it's worth, it's precisely [what we do at Verso](https://verso-optim.com/route-optimization/) in order to fund our work on this open-source project.

**Comment by AhtashamImtiyaz** [comment]
Hi jcoupey and nilsnolde, Thanks for the suggestions. We will look into the verso. Currently, I am using google distance matrix api for creating the matrices and i will going to use it with vroom. If anybody else want to do the same, then they can find it here

https://developers.google.com/optimization/routing/vrp#distance_matrix_api


I just want to know one thing, vrooms use duration matrix or distance matrix?

**Comment by jcoupey** [comment]
From the docs:

> A matrix object is an array of arrays of unsigned integers describing the rows of a custom travel-time matrix

**Comment by AhtashamImtiyaz** [comment]
Yes. I have set durations matrix in the vroom to solve it, but the results were strange. Maybe i am doing something wrong.

**Comment by jcoupey** [comment]
Closing here as the initial question seems to have been answered.

**Comment by AhtashamImtiyaz** [comment]
Hi jcoupey,

I have successfully integrated google maps with vroom. But i found there is no distance when we feed matrix to vroom. I also have distance matrix as well. How to feed that to vroom as well? Is there a way?

**Comment by jcoupey** [comment]
When using one of the available routing engines (currently OSRM or ORS), what happens sequentially is:

1. a matrix request is fired to get all required travel times
2. the optimization problem is solved
3. if extra information is required with the `-g` flag, then a "simple" routing request is sent to the routing engine in order to populate the response with distances and route geometry.

Step 3 is optional and happens only after all routes are decided so we don't actually use (or need) the whole distance matrix for this.

Now when you feed VROOM your own matrix, you're bypassing step 1. You could then technically still go through step 3, but only using a supported routing engine.

If you want to make use of the distance matrix you have, you could either:

- use the steps ordering in the solution to recompute the distance indicators yourself;
- embed the request handling within VROOM.

This last option would amount to add a new routing wrapper for the google API alongside the current OSRM/ORS wrappers.

**Comment by AhtashamImtiyaz** [comment]
> When using one of the available routing engines (currently OSRM or ORS), what happens sequentially is:
> 
> 1. a matrix request is fired to get all required travel times
> 2. the optimization problem is solved
> 3. if extra information is required with the `-g` flag, then a "simple" routing request is sent to the routing engine in order to populate the response with distances and route geometry.
> 
> Step 3 is optional and happens only after all routes are decided so we don't actually use (or need) the whole distance matrix for this.
> 
> Now when you feed VROOM your own matrix, you're bypassing step 1. You could then technically still go through step 3, but only using a supported routing engine.
> 
> If you want to make use of the distance matrix you have, you could either:
> 
> * use the steps ordering in the solution to recompute the distance indicators yourself;
> * embed the request handling within VROOM.
> 
> This last option would amount to add a new routing wrapper for the google API alongside the current OSRM/ORS wrappers.


Adding google api alongside the current OSRM and ORS is a good idea but i managed to solve it. In my case, i am sending 2 requests to the vroom solver, in first request, i am sending time matrix to get the duration and it is perfect. In another request, i am replace time matrix with distance matrix (i am extracting this from google distance matrix api). And vroom is returning the distance as well along in the cost which is also correct. This is a workaround. I got the idea about this when i found that vroom has distance when it makes a request to osrm.

**Comment by jcoupey** [comment]
You need to be aware that if you do this, the **distances** passed in input will be used for **timing** validity checks against vehicle/jobs time windows (if any).

But most importantly, there is totally no guarantee that the solutions you'll get will be the same, so the distances from the second request may not match at all with the routes from the first.

**Comment by AhtashamImtiyaz** [comment]
> You need to be aware that if you do this, the **distances** passed in input will be used for **timing** validity checks against vehicle/jobs time windows (if any).
> 
> But most importantly, there is totally no guarantee that the solutions you'll get will be the same, so the distances from the second request may not match at all with the routes from the first.

Hey mate,

Yeah i understand that. In my case, i am not using the time window. As i want only one driver. And it is working fine, i have tested it with routing on google maps and it matched

**Comment by jfalameda** [comment]
@AhtashamImtiyaz Did you find a workaround to solve matrices bigger than 25 items? I am getting a limit error from google, thus I need to compute the matrix on chunks.

**Comment by jcoupey** [comment]
@jfalameda troubleshooting API errors from an external service that we don't even integrate is beyond the scope of this bugtracker. Please turn to dedicated google-related help/support forum or channels.

**Comment by AhtashamImtiyaz** [comment]
> @AhtashamImtiyaz Did you find a workaround to solve matrices bigger than 25 items? I am getting a limit error from google, thus I need to compute the matrix on chunks.

Because this is related to feeding matrix to vroom, i am sharing this link that will help you how matrices more than 25 can be solved. Here is the details

https://developers.google.com/optimization/routing/vrp#api-requests

After you get matrices, merge the arrays and then feed the array of matrices to vroom t get the desire results.

Best of luck.


And @jcoupey thank you so much for this amazing solution.

**Comment by AhtashamImtiyaz** [comment]
Hi @jcoupey,

I tried to debug vroom and found that all the drivers and jobs locations are being sent to the table service in osrm which returns a matrix. My question is, if one of the driver's profile is not a car but a bike. Does vroom send 2 requests separately in that case? What happens in this case?

**Comment by jcoupey** [comment]
@AhtashamImtiyaz this is out of scope too for this (already closed) issue because your question has nothing to do with the original subject. Please don't keep old tickets alive with new content because it's highly confusing for anyone trying to lookup information in the bugtracker. The best way is to:

- try getting an answer from the docs;
- lookup existing issues that may cover your question (spoiler: there is one);
- open a new ticket for a new dedicated subject if required.
